src/rules/spaceBeforeFunctionParenRule.ts

```diff
@@ -0,0 +1,183 @@
+import * as ts from "typescript";
+import * as Lint from "../index";
+
+const ALWAYS_OR_NEVER = {
+    enum: ["always", "never"],
+    type: "string",
+};
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static metadata: Lint.IRuleMetadata = {
+        description: "Require or disallow a space before function parenthesis",
+        optionExamples: [
+            `true`,
+            `[true, "always"]`,
+            `[true, "never"]`,
+            `[true, {"anonymous": "always", "named": "never", "asyncArrow": "always"}]`,
+        ],
+        options: {
+            properties: {
+                anonymous: ALWAYS_OR_NEVER,
+                asyncArrow: ALWAYS_OR_NEVER,
+                constructor: ALWAYS_OR_NEVER,
+                method: ALWAYS_OR_NEVER,
+                named: ALWAYS_OR_NEVER,
+            },
+            type: "object",
+        },
+        optionsDescription: Lint.Utils.dedent`
+            One argument which is an object which may contain the keys \`anonymous\`, \`named\`, and \`asyncArrow\`
+            These should be set to either \`"always"\` or \`"never"\`.
+
+            * \`"anonymous"\` checks before the opening paren in anonymous functions
+            * \`"named"\` checks before the opening paren in named functions
+            * \`"asyncArrow"\` checks before the opening paren in async arrow functions
+            * \`"method"\` checks before the opening paren in class methods
+            * \`"constructor"\` checks before the opening paren in class constructors
+        `,
+        ruleName: "space-before-function-paren",
+        type: "style",
+        typescriptOnly: false,
+    };
+    public static INVALID_WHITESPACE_ERROR = "Spaces before function parens are disallowed";
+    public static MISSING_WHITESPACE_ERROR = "Missing whitespace before function parens";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new FunctionWalker(sourceFile, this.getOptions()));
+    }
+}
+
+class FunctionWalker extends Lint.RuleWalker {
+    private scanner: ts.Scanner;
+    private cachedOptions: any;
```

Comment:
Oh thank you. I kept using the string values in so that it allows the user to ignore certain rules (as opposed to forcing a failure if they are not consistent).